airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/dataflow/state/stats/StateAdditionalStatsStore.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.dataflow.state.stats
+
+import io.airbyte.cdk.load.command.DestinationStream
+import io.airbyte.cdk.load.dataflow.state.AdditionalStatsHistogram
+import io.airbyte.cdk.load.dataflow.state.Histogram
+import io.airbyte.cdk.load.dataflow.state.PartitionKey
+import jakarta.inject.Singleton
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * A singleton class responsible for managing additional state statistics, specifically histograms
+ * of metrics associated with partitions and destination streams. This class provides methods to add
+ * metrics values, retrieve combined statistics, and maintain metrics completeness with default
+ * values.
+ */
+@Singleton
+class StateAdditionalStatsStore {
+
+    private val store =
+        ConcurrentHashMap<
+            PartitionKey, ConcurrentHashMap<DestinationStream.Descriptor, AdditionalStatsHistogram>
+        >()
+
+    /**
+     * Adds a value to the histogram associated with the given partition and stream descriptor for
+     * the specified metric.
+     *
+     * @param partitionKey The key representing the partition for which the histogram is maintained.
+     * @param streamDescriptor The descriptor identifying the specific destination stream.
+     * @param metric The metric whose value is to be incremented in the histogram.
+     * @param value The value to increment for the specified metric.
+     */
+    fun add(
+        partitionKey: PartitionKey,
+        streamDescriptor: DestinationStream.Descriptor,
+        metric: ObservabilityMetrics,
+        value: Double
+    ) {
+        store.putIfAbsent(partitionKey, ConcurrentHashMap())
+        store[partitionKey]?.let { partitionStats ->
+            partitionStats.putIfAbsent(streamDescriptor, Histogram())
+            partitionStats[streamDescriptor]?.increment(metric.metricName, value)
+        }
+    }
+
+    /**
+     * Drains the accumulated statistics for the specified destination stream across the provided
+     * partition keys, returning a combined histogram of all relevant metrics. For any metric values
+     * not present, default values are added. This is to ensure that the metrics are always
+     * published as part of the state message with a zero value if no value has been recorded since
+     * the last state message publish event.
+     *
+     * @param partitionKeys The list of partition keys to process, each representing a distinct
+     * partition.
+     * @param stream The descriptor of the destination stream whose statistics are to be drained.
+     * @return A `Histogram` containing the combined metrics for the specified destination stream.
+     */
+    fun drain(
+        partitionKeys: List<PartitionKey>,
+        stream: DestinationStream.Descriptor
+    ): Histogram<String> {
+        return partitionKeys
+            .mapNotNull { store[it] }
```

Comment:
If we don't remove the histo it's a small memory leak

Reply:
This isn't the histogram.  This is the map of stream -> histogram.  The histogram gets removed on line 70.